### PR TITLE
use expected/actual style for assert and add stats

### PR DESCRIPTION
**Pull request by soelgary**
This is by no means complete, but it is an improvement from before. The main difference is that this change adds some stats to let us know how many tests passed and failed. Next step is to print the expected and actual for failed tests, but that is for another day

### CODE REVIEW

File: src/test.sig [hunk 1]
```diff
@@ -1,7 +1,10 @@
 signature TEST =
 sig
   exception Failed
-
+  val testsRan : int ref
+  val testsPassed : int ref
   val test : (unit -> bool) -> bool
-  val assert : bool -> bool
+  val assert : ''a * ''a -> bool
```
Comment (nixpulvis) [inline src/test.sig]: what are double `'` parameters?

File: src/test.sig [hunk 2]
```diff
@@ -1,7 +1,10 @@
 signature TEST =
 sig
   exception Failed
-
+  val testsRan : int ref
+  val testsPassed : int ref
   val test : (unit -> bool) -> bool
-  val assert : bool -> bool
```
Comment (nixpulvis) [inline src/test.sig:6]: I propose leaving an `assert : bool -> bool` and adding the `assertEq` function.

File: src/test.sml [hunk 3]
```diff
@@ -1,7 +1,23 @@
...
   exception Failed
 
   fun test f = f () handle Failed => false
-  fun assert b = if b then true else raise Failed
+
+  fun assert(expected,actual) =
+    (testsRan := !testsRan + 1;
+     if expected = actual
+     then (testsPassed := !testsPassed + 1; true)
+     else raise Failed)
```
Comment (nixpulvis) [inline src/test.sml]: I'm still new to exceptions in SML but can we pass along the data for the failure here?

Comment (nixpulvis) [inline src/test.sml]: Also considering we are using exceptions here, the return type of `bool` is pretty meaningless. We could get away with a unit return type.

File: src/test.sig (hunk 1 above)
Comment (soelgary) [inline src/test.sig]: It means the type implements equals so we can say `expected = actual`

File: src/test.sig (hunk 2 above)
Comment (soelgary) [inline src/test.sig:6]: fair enough

File: src/test.sig (hunk 1 above)
Comment (nixpulvis) [inline src/test.sig]: Oh really, how weird. I hate how SML special cases so many things like this.

File: src/test.sml (hunk 3 above)
Comment (soelgary) [inline src/test.sml]: Not sure about how to pass data along. It isn't needed for this so Ill save it for another day.

Fair enough.

Comment (nixpulvis) [inline src/test.sml]: Sounds good.

**Comment by nixpulvis** [comment]
Running the tests I get:

```
[opening test.sml]
[autoloading]
src/test.sml:13.18 Warning: calling polyEqual
```

**Comment by nixpulvis** [comment]
And the error mysteriously disappeared. Color me confused, but this looks good anyway.

**Comment by nixpulvis** [comment]
I've pushed complete tests for the 7 string cases in `fixtures`. As you can see when running the tests the position isn't correctly tracking the source program. Strings like `"This has an \062 ascii char"` look at the length of the output string, not the input string as they should.

**Comment by nixpulvis** [comment]
That said I'm going to open a new issue, and merge this PR.

**Pull request merged**